### DISCUSSION TITLE
PyPI Package (pip) Command Line tool

### DISCUSSION THREAD

**Discussion by Saketh-Chandra** [ThioJoe/YT-Spammer-Purge#42] [Ideas]
We can make a [PyPI package](https://pypi.org/) So that people can install it.
- like: `pip install YouTube-Spammer-Purge`.
- After installing the PIP (PyPI ) package
   -  simply people can type `YouTube-Spammer-Purge` in the terminal/CMD And the script will start.
   
If we do that, people can't get a Browser download warning or windows exe running warning.

I have some experience in making PIP in making python libraries.
- `Helloworldpy` is my simple PyPi (pip) package. we can run it through terminal [Source Code ](https://github.com/Saketh-Chandra/helloworldpy), [PyPI package link](https://pypi.org/project/helloworldpy/)

Can I make ` Pull Requests`? If implemented that feature?

**Comment by ThioJoe** [comment]
I'll just look into implementing it, I'd rather be able to learn to do it myself

**Comment by Saketh-Chandra** [comment]
Yeah fine!

**Comment by ryleu** [comment]
Setting up a pypi package would make cross-platform installation much easier. I would also recommend setting up a poetry toml file (make sure to add `poetry.lock` to your `.gitignore`!) to make poetry venv installs easier.

**Reply by rachmadaniHaryono** [reply]
i create #459 for this

**Comment by KendallDoesCoding** [comment]
> We can make a [PyPI package](https://pypi.org/) So that people can install it.
> 
> * like: `pip install YouTube-Spammer-Purge`.
> * After installing the PIP (PyPI ) package
>   
>   * simply people can type `YouTube-Spammer-Purge` in the terminal/CMD And the script will start.
> 
> If we do that, people can't get a Browser download warning or windows exe running warning.
> 
> I have some experience in making PIP in making python libraries.
> 
> * `Helloworldpy` is my simple PyPi (pip) package. we can run it through terminal [Source Code ](https://github.com/Saketh-Chandra/helloworldpy), [PyPI package link](https://pypi.org/project/helloworldpy/)
> 
> Can I make ` Pull Requests`? If implemented that feature?

Seems interesting..

**Comment by KendallDoesCoding** [comment]
This may help: https://towardsdatascience.com/how-to-publish-a-python-package-to-pypi-7be9dd5d6dcd

**Comment by rachmadaniHaryono** [comment]
## pr for poetry

so i create #459 and it is recently closed

reason https://github.com/ThioJoe/YT-Spammer-Purge/pull/459#issuecomment-1030466885

> Still can't figure out why this is needed and no one has explained it, let alone in the initial PR

so

https://github.com/ThioJoe/YT-Spammer-Purge/discussions/42#discussioncomment-1597192

> I'll just look into implementing it, I'd rather be able to learn to do it myself

based on comment above, i assume thiojoe would use other packaging setup like setup.py or pdm

i don't know about pdm vs poetry but these article may help choosing poetry over setup.py

https://nanthony007.medium.com/stop-using-pip-use-poetry-instead-db7164f4fc72

https://hackersandslackers.com/python-poetry-package-manager/

## discussion and packaging 

this could help following discussion

https://github.com/ThioJoe/YT-Spammer-Purge/discussions/541

https://github.com/ThioJoe/YT-Spammer-Purge/discussions/515

https://github.com/ThioJoe/YT-Spammer-Purge/discussions/506

https://github.com/ThioJoe/YT-Spammer-Purge/discussions/405

https://github.com/ThioJoe/YT-Spammer-Purge/discussions/213

https://github.com/ThioJoe/YT-Spammer-Purge/discussions/281

https://github.com/ThioJoe/YT-Spammer-Purge/discussions/315

https://github.com/ThioJoe/YT-Spammer-Purge/discussions/258

https://github.com/ThioJoe/YT-Spammer-Purge/discussions/208

https://github.com/ThioJoe/YT-Spammer-Purge/discussions/144

https://github.com/ThioJoe/YT-Spammer-Purge/discussions/54

https://github.com/ThioJoe/YT-Spammer-Purge/discussions/65

https://github.com/ThioJoe/YT-Spammer-Purge/discussions/64


## entry point

note: this section taken from pr discussion

i try to add entry point like this [python-poetry.org/docs/pyproject/#scripts](https://python-poetry.org/docs/pyproject/#scripts)

but it require to run a function

but in YTSpammerPurge.py it is hidden under `if __name__ == '__main__'` section

e: if i try to add

```toml
[tool.poetry.scripts]
YTSpammerPurge = 'YTSpammerPurge'
```

it will raise error when run with poetry

```python
> poetry run YTSpammerPurge

  ValueError

  not enough values to unpack (expected 2, got 1)

  at ~/.local/pipx/venvs/poetry/lib/python3.9/site-packages/poetry/console/commands/run.py:36 in run_script
      32│     def run_script(self, script, args):
      33│         if isinstance(script, dict):
      34│             script = script["callable"]
      35│
    → 36│         module, callable_ = script.split(":")
      37│
      38│         src_in_sys_path = "sys.path.append('src'); " if self._module.is_in_src() else ""
      39│
      40│         cmd = ["python", "-c"]
```

recommendation

```diff
diff --git a/YTSpammerPurge.py b/YTSpammerPurge.py
index 79ef3de..e30fd74 100644
--- a/YTSpammerPurge.py
+++ b/YTSpammerPurge.py
@@ -1494,7 +1494,7 @@ def main():


 # Runs the program
-if __name__ == "__main__":
+def run_main()
 #   #For speed testing

   # import cProfile
@@ -1585,3 +1585,5 @@ if __name__ == "__main__":
   else:
     print("\nFinished Executing.")

+if __name__ == '__main__':
+    run_main()
```

this way i can put

```toml
[tool.poetry.scripts]
YTSpammerPurge = 'YTSpammerPurge:run_main'
```

it is still not enough for packaging but can be used in the future

## current structure

now the program assume all asset/config in current directory so it harder to package it

imo the program need cli parameter or some input and have user set the option which file go where

**Reply by ryleu** [reply]
It should use environment variables if a file named config.json (or similar) is not in the current working directory. That is what I’ve done in the past for a few other projects and it works wonderfully.

**Reply by rachmadaniHaryono** [reply]
https://github.com/ThioJoe/YT-Spammer-Purge/blob/1c50d30ea5d78a92147ec4dfbb4cb2210a0262dd/Scripts/auth.py#L42

maybe something like this

```python
# ...
import os
# ...
def get_authenticated_service():
   # ...
  CLIENT_SECRETS_FILE = os.getenv('YT_SPAMMER_PURGE_CLIENT_SECRETS_FILE', 'client_secrets.json')
  #  ...
```

**Reply by rachmadaniHaryono** [reply]
this is example when uploading to pypi

https://test.pypi.org/project/yt-spammer-purge/

it doesn't mean yt-spammer-purge is working when installing from pypi

it just mean pypi upload is succesful

upload is based from this guide https://stackoverflow.com/a/69026451

```shell
> tar tf YT-Spammer-Purge-2.16.0rc4.tar.gz | tree --fromfile .
.
└── YT-Spammer-Purge-2.16.0rc4
    ├── LICENSE
    ├── PKG-INFO
    ├── README.md
    ├── Scripts
    │   ├── auth.py
    │   ├── community_downloader.py
    │   ├── files.py
    │   ├── gui.py
    │   ├── logging.py
    │   ├── operations.py
    │   ├── prepare_modes.py
    │   ├── shared_imports.py
    │   ├── utils.py
    │   └── validation.py
    ├── YTSpammerPurge.py
    ├── assets
    │   ├── SpamAccountsList.txt
    │   ├── SpamDomainsList.txt
    │   ├── SpamThreadsList.txt
    │   ├── SpamVersionInfo.json
    │   ├── cancelButton.png
    │   ├── default_config.ini
    │   ├── executeButton.png
    │   ├── icon.png
    │   ├── inputSubmitButton.png
    │   ├── inputTextBox.png
    │   ├── outputTextBox.png
    │   └── rootZoneDomainList.txt
    ├── pyproject.toml
    └── setup.py
```
